File: app/src/main/java/io/github/wulkanowy/ui/main/timetable/TimetablePresenter.java
package io.github.wulkanowy.ui.main.timetable;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.github.wulkanowy.data.RepositoryContract;
import io.github.wulkanowy.ui.base.BasePresenter;
import io.github.wulkanowy.utils.TimeUtils;

public class TimetablePresenter extends BasePresenter<TimetableContract.View>
        implements TimetableContract.Presenter {

    private List<String> dates = new ArrayList<>();

    private int positionToScroll;

    private boolean isFirstSight = false;

    @Inject
    TimetablePresenter(RepositoryContract repository) {
        super(repository);
    }

    @Override
    public void onStart(TimetableContract.View view, boolean primary) {
        super.onStart(view);

        if (dates.isEmpty()) {
            dates = TimeUtils.getMondaysFromCurrentSchoolYear();
        }
        positionToScroll = dates.indexOf(TimeUtils.getDateOfCurrentMonday());

        if (primary) {
            onFragmentVisible(true);
        }
    }

    @Override
    public void onFragmentVisible(boolean isSelected) {
        if (isSelected) {
            getView().setActivityTitle();

            if (!isFirstSight) {
                isFirstSight = true;

                for (String date : dates) {
                    getView().addPageToAdapter(TimetableTabFragment.newInstance(date), date);
                }
                getView().setAdapterWithTabLayout();
                getView().scrollViewPagerToPosition(positionToScroll);
            }
        }
    }

    @Override
    public void onTabSelected(int position) {
        getView().setChildFragmentSelected(position, true);
    }

    @Override
    public void onTabUnselected(int position) {
        getView().setChildFragmentSelected(position, false);
    }
}

File: app/src/main/java/io/github/wulkanowy/ui/main/timetable/TimetablePresenter.kt
package io.github.wulkanowy.ui.main.timetable

import eu.davidea.flexibleadapter.items.AbstractFlexibleItem
import io.github.wulkanowy.data.ErrorHandler
import io.github.wulkanowy.data.db.entities.Semester
import io.github.wulkanowy.data.db.entities.Timetable
import io.github.wulkanowy.data.repositories.SessionRepository
import io.github.wulkanowy.data.repositories.TimetableRepository
import io.github.wulkanowy.ui.base.BasePresenter
import io.github.wulkanowy.utils.extension.isHolidays
import io.github.wulkanowy.utils.extension.toFormat
import io.github.wulkanowy.utils.getNearMonday
import io.github.wulkanowy.utils.schedulers.SchedulersManager
import org.threeten.bp.LocalDate
import java.util.*
import javax.inject.Inject

class TimetablePresenter @Inject constructor(
        private val errorHandler: ErrorHandler,
        private val schedulers: SchedulersManager,
        private val timetableRepository: TimetableRepository,
        private val sessionRepository: SessionRepository
) : BasePresenter<TimetableView>(errorHandler) {

    var currentDate: LocalDate = getNearMonday(LocalDate.now())
        private set

    override fun attachView(view: TimetableView) {
        super.attachView(view)
        view.initView()
    }

    fun loadTimetableForPreviousWeek() = loadData(currentDate.minusDays(7).toEpochDay())

    fun loadTimetableForNextWeek() = loadData(currentDate.plusDays(7).toEpochDay())

    fun loadData(date: Long?, forceRefresh: Boolean = false) {
        this.currentDate = LocalDate.ofEpochDay(date ?: getNearMonday(currentDate).toEpochDay())
        if (currentDate.isHolidays()) return

        disposable.clear()
        disposable.add(sessionRepository.getSemesters()
                .map { selectSemester(it, -1) }
                .flatMap { timetableRepository.getTimetable(it, currentDate, forceRefresh) }
                .map { it.groupBy { exam -> exam.date }.toSortedMap() }
                .map { createTimetableItems(it) }
                .subscribeOn(schedulers.backgroundThread())
                .observeOn(schedulers.mainThread())
                .doOnSubscribe {
                    view?.run {
                        showRefresh(forceRefresh)
                        showProgress(!forceRefresh)
                        if (!forceRefresh) showEmpty(false)
                        showContent(null == date && forceRefresh)
                        showPreButton(!currentDate.minusDays(7).isHolidays())
                        showNextButton(!currentDate.plusDays(7).isHolidays())
                        updateNavigationWeek("${currentDate.toFormat("dd.MM")}-${currentDate.plusDays(4).toFormat("dd.MM")}")
                    }
                }
                .doAfterSuccess {
                    view?.run {
                        showEmpty(it.isEmpty())
                        showContent(it.isNotEmpty())
                    }
                }
                .doFinally {
                    view?.run {
                        showRefresh(false)
                        showProgress(false)
                    }
                }
                .subscribe({ view?.updateData(it) }) { errorHandler.proceed(it) })
    }

    private fun createTimetableItems(items: Map<Date, List<Timetable>>): List<TimetableHeader> {
        return items.map {
            TimetableHeader().apply {
                date = it.key
                subItems = it.value.map { item ->
                    TimetableItem().apply { lesson = item }
                }
            }
        }
    }

    fun onTimetableItemSelected(item: AbstractFlexibleItem<*>?) {
        if (item is TimetableItem) view?.showTimetableDialog(item.lesson)
    }

    private fun selectSemester(semesters: List<Semester>, index: Int): Semester {
        return semesters.single { it.current }.let { currentSemester ->
            if (index == -1) currentSemester
            else semesters.single { semester ->
                semester.run {
                    semesterName - 1 == index && diaryId == currentSemester.diaryId
                }
            }
        }
    }
}
